bgpd/bgp_conditional_adv.h

```diff
@@ -0,0 +1,130 @@
+/*
+ * BGP Conditional advertisement
+ * Copyright (C) 2020  Samsung Research Institute Bangalore.
+ * 		       Madhurilatha Kuruganti
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef _FRR_BGP_CONDITION_ADV_H
+#define _FRR_BGP_CONDITION_ADV_H
+#include <zebra.h>
+#include "prefix.h"
+#include "bgpd/bgp_addpath.h"
+#include "bgpd/bgp_attr.h"
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_debug.h"
+#include "bgpd/bgp_route.h"
+#include "bgpd/bgp_updgrp.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Polling time for monitoring condition-map routes in route table */
+#define CONDITIONAL_ROUTES_POLL_TIME 60
+
+#define FOREACH_ACCESS_LIST_FILTER(alist, filter)                              \
+	for (filter = alist->head; filter; filter = filter->next)
+
+static inline bool is_rmap_valid(struct route_map *rmap)
+{
+	if (!rmap || !rmap->head)
+		return false;
+
+	/* Doesn't make sense to configure advertise
+	 * or condition map in deny/any clause. */
+	if (!rmap->head->type == RMAP_PERMIT)
+		return false;
+
+	/* If a match command is not present, all routes match the clause */
+	if (!rmap->head->match_list.head)
+		return false;
+
+	return true;
+}
+
+static inline afi_t get_afi_from_str(const char *str)
+{
+	if (!strcmp(str, "ip address"))
+		return AFI_IP;
+	else if (!strcmp(str, "ipv6 address"))
+		return AFI_IP6;
+	else
+		return AFI_MAX;
+}
+
+static inline bool advertise_dest_routes(struct update_subgroup *subgrp,
```

Comment:
> 
> 
> This function seems a bit big to be inlined, is there any specific reason?

We are calling this function for each bgp_dest node in the table.
if the Table size is big there could be more number of function calls thats why. and we poll the table every 60seconds once. If you feel its ok to remove inline. will do that.

Reply:
> 
> 
> I have another one question regarding this function: can't `subgroup_announce_table` be adopted for all the changes we do here? I mean you introduced a few functions which basically do almost the same with some additional conditions.

for some cases subgroup_announce_table() is used. but not for all to avoid unnecessary checks and look-up
example route-map attached to advertise-map is removed but previously some routes were withdrawn.
in that case subgroup_announce_table() is used to send normal update.

Reply:
this function is removed after adding code to support other match rules.